### PR TITLE
[MAINTENANCE] Run CLI tests as a separate stage in Azure pipelines

### PR DESCRIPTION
**Pull request by alexsherstinsky**
### Rationale
* CLI tests use `CliRunner`, which invokes a system call to execute the CLI command.  Other tests do not incur inter-process communication.  Thus, it is advantageous to isolate CLI tests in Azure pipelines to reduce test execution time (see below).
* CLI tests do not need to go through the matrix of backend choices (Pandas, Spark, and the SQL Databases), because CLI relies on `DataContext` exclusively -- and `DataContext` APIs are thoroughly tested as part of the general test suite.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

File: azure-pipelines.yml
Comment (talagluck) [inline azure-pipelines.yml]: Just checking - is it ok to update pip now because we have the new constraints-dev file and that gets us around dependency issues?

File: great_expectations/profile/user_configurable_profiler.py
Comment (talagluck) [inline great_expectations/profile/user_configurable_profiler.py]: Why was this added?

File: tests/profile/test_user_configurable_profiler_v3_batch_request.py
Comment (talagluck) [inline tests/profile/test_user_configurable_profiler_v3_batch_request.py]: Makes sense!

**Review by talagluck — COMMENTED** [review]
Looks good! Just a couple of questions. I'll wait for someone with a bit more Azure experience to approve.

File: great_expectations/profile/user_configurable_profiler.py
Comment (alexsherstinsky) [reply great_expectations/profile/user_configurable_profiler.py]: To prevent warnings about using private methods.

File: azure-pipelines.yml
Comment (alexsherstinsky) [reply azure-pipelines.yml]: This is under test.  If does not work, will revert.

File: azure-pipelines.yml
Comment (anthonyburdi) [inline azure-pipelines.yml:329]: Could this maybe just be dependent on `[scope_check, lint]` so that it can run in parallel to the other stages? To speed up CI testing.
```suggestion
    dependsOn: [scope_check, lint]
```

**Review by anthonyburdi — APPROVED** [review]
Alex - this is awesome! Just one non-blocking suggestion. If you don't agree leave it as is. I'll approve pending CI passing.

File: azure-pipelines.yml
Comment (alexsherstinsky) [reply azure-pipelines.yml:329]: @anthonyburdi This is a good question -- it was done on purpose.  Here is the bulleted reasoning:
* CLI tests are in the category of integration tests, because CLI functions utilize the lower-level (`DataContext`) API.
* To save time, integration tests do not exercise every backend as long as the tests for the underlying platform API perform all the required tests for backends, different versions of `Python`, `pandas`, etc.
* Therefore, to insure that the CLI integration tests run on top of a platform, fully tested on a broad coverage of tests, the above dependency is created.

Thanks!

File: great_expectations/dataset/sqlalchemy_dataset.py
Comment (Shinnnyshinshin) [inline great_expectations/dataset/sqlalchemy_dataset.py:65]: Thank you for this :)

File: setup.py
Comment (Shinnnyshinshin) [inline setup.py]: If we are requiring sqlalchemy as an `extra_require`,  will need to make it `>=1.4.7` as it includes the bugfix we submitted.

File: azure-pipelines.yml
Comment (Shinnnyshinshin) [inline azure-pipelines.yml:367]: whoohoo!

**Review by Shinnnyshinshin — DISMISSED** [review]
just a small request for Sqlalchemy version >= 1.4.7

File: setup.py
Comment (alexsherstinsky) [reply setup.py]: @Shinnnyshinshin Do we need to do the same in `requirements-dev-sql.txt`?  Or only here in `setup.py`?  Thanks!

File: azure-pipelines.yml
Comment (jcampbell) [inline azure-pipelines.yml:358]: Don't we have support for pyspark 3 now? Can we change requirements-dev?

**Review by jcampbell — APPROVED** [review]
Small tweak note

File: azure-pipelines.yml
Comment (anthonyburdi) [reply azure-pipelines.yml:329]: @alexsherstinsky - All of that totally makes sense. Though we do only allow PRs to merge that have passed all of the required/comprehensive/backend/compatibility checks anyway. So if we run this integration stage in parallel then we only risk wasting the CI resources to run this integration if it would have been pre-empted by an earlier failure -- not that other items would be untested. I don't currently know the economics / parallel configuration of our CI pipeline, but faster builds for expensive developers at the expense of more cheap robots sounds like a good tradeoff to me.

Comment (anthonyburdi) [reply azure-pipelines.yml:329]: This is a non-blocking comment though - I just wanted to raise a potential developer efficiency opportunity here 😄

File: azure-pipelines.yml
Comment (alexsherstinsky) [reply azure-pipelines.yml:358]: @jcampbell I was *hoping* that we could change `requirements-dev-spark.txt` to use `Spark 3.x`.  However, I was a little uneasy about it for three reasons:
* Has everybody on the Engineering Team at Superconductive already standardized on `Spark 3.x`?  And does it work smoothly for everybody in terms of S3 access and any other relevant integrations?
* Are there companies still using `Spark 2.4`?  I recall that as of a year ago, companies were upgrading from `Spark 2.3` to `Spark 2.4` and not yet on `Spark 3.x`. 
* IMPORTANT: `Spark 3.x` does *NOT* support `Python <3.8` -- this might present difficulties to companies using `Python 3.7` and `Python 3.6` (we validate those `Python` versions in compatibility testing in Azure). 

On the basis of this reasoning, I felt that it would be best to defer the global requirement change to a future task.

Thank you.  /cc @eugmandel

File: setup.py
Comment (Shinnnyshinshin) [reply setup.py]: @alexsherstinsky I believe it should be both

Comment (alexsherstinsky) [reply setup.py]: @Shinnnyshinshin I just committed the change -- please review carefully and confirm by approving that it is good.  Thanks so much!

File: azure-pipelines.yml
Comment (alexsherstinsky) [reply azure-pipelines.yml:329]: 🙇

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you very much for this :) LGTM

File: setup.py
Comment (alexsherstinsky) [reply setup.py]: @Shinnnyshinshin Please double-check based on our latest discussion.

**Pull request merged**
